src/types.rs

```diff
@@ -368,6 +372,57 @@ impl<'b> From<&'b mut crate::ffi::QUIC_STREAM_EVENT> for StreamEvent<'b> {
     }
 }
 
+/// Buffer with same abi as ffi type.
+/// # Safety
+/// It has no ownership of the memory chunk,
+/// and user needs to ensure that this ref has
+/// the same lifetime as the original buffer
+/// location.
+#[repr(transparent)]
+pub struct BufferRef(pub QUIC_BUFFER);
+
+impl BufferRef {
+    /// Get the bytes of the buffer.
+    pub fn as_bytes(&self) -> &[u8] {
+        unsafe { slice_conv(self.0.Buffer, self.0.Length as usize) }
```

Comment:
I come after the PR is merged, but isn't this breaking rust safety?
When exiting the unsafe block, nothing guarantees the slice refers to valid memory.

I don't know how we could express it, but as far as I understand, this breaks Rust memory safety guarantees.

Reply:
Returning the slice is not unsafe. std Vec does the same thing.
What is unsafe is constructing the BufferRef type in the first place, since the lifetime info of the buffer is not captured by BufferRef struct. We could consider making all the constructing function unsafe.

Reply:
Good point, I should have put that comment on the constructing functions below.

I think we should make the constructing functions unsafe for now - they break safe Rust code guarantees; we should make sure a user of the lib is aware of it.

We can potentially make a smarter wrapper later on that could capture the lifetime one way or another. I am currently thinking the receive completion function could take BufferRefs as a parameter to consume them. It would likely take some extra sate to ensure receives are completed in order though, and that it respects the rules for the different receive modes (multiple receives vs single).